lib/event.c

```diff
@@ -572,13 +639,58 @@ struct event_loop *event_master_create(const char *name)
 	set_nonblocking(rv->io_pipe[0]);
 	set_nonblocking(rv->io_pipe[1]);
 
+#if defined(USE_EPOLL) &&                                                      \
+	(defined(HAVE_EPOLL_WAIT) || defined(HAVE_EPOLL_PWAIT) ||              \
+	 defined(HAVE_EPOLL_PWAIT2))
+	rv->awakened = false;
+
+	/* Initialize data structures for epoll() */
+	rv->handler.event_size = rv->fd_limit;
+	rv->handler.event_count = 0;
+	rv->handler.new_event_count = 0;
+	rv->handler.regular_revent_count = 0;
+	rv->handler.epoll_fd = epoll_create1(0);
+	rv->handler.events =
+		XCALLOC(MTYPE_EVENT_MASTER, sizeof(struct frr_epoll_event) *
+						    rv->handler.event_size);
+	rv->handler.event_ptrs =
+		XCALLOC(MTYPE_EVENT_MASTER, sizeof(struct frr_epoll_event *) *
+						    rv->handler.event_size);
+	rv->handler.new_events =
+		XCALLOC(MTYPE_EVENT_MASTER, sizeof(struct frr_epoll_event) *
+						    rv->handler.event_size);
+	rv->handler.new_event_ptrs =
+		XCALLOC(MTYPE_EVENT_MASTER, sizeof(struct frr_epoll_event *) *
+						    rv->handler.event_size);
+	rv->handler.revents =
+		XCALLOC(MTYPE_EVENT_MASTER,
+			sizeof(struct epoll_event) * rv->handler.event_size);
+	rv->handler.regular_revents =
+		XCALLOC(MTYPE_EVENT_MASTER,
+			sizeof(struct epoll_event) * rv->handler.event_size);
+	for (i = 0; i < rv->handler.event_size; i++) {
+		rv->handler.events[i].index = -1;
+		rv->handler.new_events[i].index = -1;
+	}
+	pipe_read_ev.data.fd = rv->io_pipe[0];
+	pipe_read_ev.events = EPOLLIN;
+	if (-1 == epoll_ctl(rv->handler.epoll_fd, EPOLL_CTL_ADD, rv->io_pipe[0],
+			    &pipe_read_ev)) {
+		flog_err(
+			EC_LIB_NO_THREAD,
+			"Attempting to call epoll_ctl to add io_pipe[0] but failed, fd: %d!",
+			rv->io_pipe[0]);
```

Comment:
What are the expected failure conditions here? It seems like if we cannot do this, this event loop is non functional. Is that accurate? If so, should we abort here?

Reply:
Failure occurs when epoll_ctl system call doesn't work normally, (e.g. fails to allocate memory or exceeds fd allocation limit). Indeed, if we encounter error here, the event loop is non-functional and we have no choice but to abort.

I realized that epoll_create1 system call above should also be wrapped with similar if statement:
from
```c
        rv->handler.epoll_fd = epoll_create1(0);
```
to
```c
	if (rv->handler.epoll_fd == -1) {
		flog_err(
			EC_LIB_NO_THREAD,
			"Couldn't create an epoll set!");
		return NULL;
	}
```

It seems that "return NULL" is not a good aborting way. I wonder if there are better ways.